### PR TITLE
Fix: pool-coordinator

### PR DESCRIPTION
**Pull request by LaurenceLiZhixin**
fix some logical bugs and nil pointer.

### CODE REVIEW

File: pkg/yurthub/storage/etcd/keycache.go [hunk 1]
```diff
@@ -120,48 +120,49 @@ func (c *componentKeyCache) Recover() error {
 }
 
 func (c *componentKeyCache) getPoolScopedKeyset() (*keySet, error) {
-	client := c.getEtcdClient()
```
Comment (Congrool) [inline pkg/yurthub/storage/etcd/keycache.go:123]: It seems the function is uninitialized. I'll fix it.

**Comment by Congrool** [comment]
BTW, could you please split your PR with several commits each of which fix one bug? It will be helpful for review, thanks!

File: pkg/yurthub/poolcoordinator/coordinator.go [hunk 2]
```diff
@@ -316,7 +311,8 @@ func (coordinator *coordinator) IsReady() (cachemanager.CacheManager, bool) {
 	// If electStatus is not PendingHub, it means pool-coordinator is healthy.
 	coordinator.Lock()
 	defer coordinator.Unlock()
-	if coordinator.electStatus != PendingHub && coordinator.isPoolCacheSynced && !coordinator.needUploadLocalCache {
+	// fixme:  coordinator.isPoolCacheSynced now is not considered
+	if coordinator.electStatus != PendingHub && !coordinator.needUploadLocalCache {
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply pkg/yurthub/poolcoordinator/coordinator.go:315]: Now isPoolCacheSynced is never set to 'true'.

File: pkg/yurthub/certificate/hubself/cert_mgr.go [hunk 3]
```diff
@@ -524,8 +524,8 @@ func createBasic(apiServerAddr string, caCert []byte) *clientcmdapi.Config {
 	}
 }
 
-// createInsecureRestClientConfig create insecure rest client config.
-func createInsecureRestClientConfig(remoteServer *url.URL) (*restclient.Config, error) {
+// CreateInsecureRestClientConfig create insecure rest client config.
+func CreateInsecureRestClientConfig(remoteServer *url.URL) (*restclient.Config, error) {
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply pkg/yurthub/certificate/hubself/cert_mgr.go:528]: I'll fix it.

File: pkg/yurthub/filter/approver_test.go [hunk 4]
```diff
@@ -43,6 +43,7 @@ var supportedResourceAndVerbsForFilter = map[string]map[string]sets.String{
 	},
 	ServiceTopologyFilterName: {
 		"endpoints":      sets.NewString("list", "watch"),
+		"pods":           sets.NewString("list", "watch"),
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply pkg/yurthub/filter/approver_test.go:46]: I'll fix it.

File: pkg/yurthub/poolcoordinator/coordinator.go [hunk 5]
```diff
@@ -203,10 +201,9 @@ func (coordinator *coordinator) Run() {
 				needUploadLocalCache = true
 				needCancelEtcdStorage = true
 				isPoolCacheSynced = false
-				etcdStorage = nil
-				poolCacheManager = nil
+				coordinator.poolCacheManager = nil
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply pkg/yurthub/poolcoordinator/coordinator.go:204]: I'll add lock here

File: cmd/yurthub/app/start.go [hunk 6]
```diff
@@ -180,40 +180,38 @@ func Run(ctx context.Context, cfg *config.YurtHubConfiguration) error {
...
 		// Waitting for the coordinator to run, before using it to create other components.
-		coordinatorHealthChecker, coordinatorTransportManager, coordinator, err = waittingForCoordinator()
+		coordinatorHealthCheckerGetter, coordinatorTransportManagerGetter, coordinatorGetter, err = coordinatorRun(ctx, cfg, restConfigMgr, cloudHealthChecker)
 		if err != nil {
 			return fmt.Errorf("failed to wait for coordinator to run, %v", err)
 		}
 	}
 
+	// wait for async coordinator informer registry
+	time.Sleep(time.Second * 5)
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply cmd/yurthub/app/start.go:200]: You are right. I'll change it to notification.

**Comment by LaurenceLiZhixin** [comment]
> BTW, could you please split your PR into several commits each of which fixes one bug? It will be helpful for review, thanks!

Sure, I'll do that later

File: pkg/yurthub/storage/etcd/keycache.go (hunk 1 above)
Comment (LaurenceLiZhixin) [reply pkg/yurthub/storage/etcd/keycache.go:123]: Will you fix it into my pr or create a new one ? @Congrool

Comment (Congrool) [reply pkg/yurthub/storage/etcd/keycache.go:123]: I'll create a new pr to fix it.

File: pkg/yurthub/poolcoordinator/coordinator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Congrool) [reply pkg/yurthub/poolcoordinator/coordinator.go:321]: Acctually we need `isPoolCacheSynced`.  I'll submit a pr to fix it.

File: pkg/yurthub/poolcoordinator/leader_election.go [hunk 7]
```diff
@@ -55,7 +55,7 @@ func NewHubElector(
 		coordinatorClient:           coordinatorClient,
 		coordinatorHealthChecker:    coordinatorHealthChecker,
 		cloudAPIServerHealthChecker: cloudAPIServerHealthyChecker,
-		electorStatus:               make(chan int32),
+		electorStatus:               make(chan int32, 1),
```
[the remark this answers is not in the document]
Comment (Congrool) [reply pkg/yurthub/poolcoordinator/leader_election.go:58]: How about moving `he.electorStatus <- PendingHub` to the begining of `Run()`?

**Review by rambohe-ch — APPROVED** [review]

File: pkg/yurthub/poolcoordinator/coordinator.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Congrool) [reply pkg/yurthub/poolcoordinator/coordinator.go:304]: fixed by #1131

File: pkg/yurthub/storage/etcd/keycache.go (hunk 1 above)
Comment (Congrool) [reply pkg/yurthub/storage/etcd/keycache.go:123]: fixed by #1137

**Pull request merged**